### PR TITLE
ipq40xx: fix reboot for compex wpj428

### PR DESCRIPTION
**Pull request by yogo1212**
Reset the mdio using dedicated GPIOs.

### CODE REVIEW

**Comment by yogo1212** [comment]
the GPIOs are documented here:
https://compex.com.sg/wp-content/uploads/2020/01/wpj428-6a04-hwg-sl-177.pdf

**Comment by chunkeey** [comment]
I'm sorry, but this makes no sense. Your patch description mentions the flash chip, but the actual changes are dealing with network / mdio code.

**Comment by yogo1212** [comment]
@chunkeey oh, oops! sorry - you're right!
i had patches prepared for the 4-byte opcode switching - that's no longer needed, luckily.
the PR description was taken from the git commit that i cherry-picked around for over half a year with the wrong text...

**Comment by chunkeey** [comment]
From what I remember, (See the commit message of the original WPJ428 commit:
2796ab85ed5097d91623e60abf22325bd9086407 ). Compex releaseed back in 2018 an important update for the WPJ428. And it fixed the reset issues. So, I wonder what's going on here? Because is this a issue that needs backporting to 19.07 as well or is this because you haven't updated. If the your commit message had more details and information about what boardrev and u-boot version you are experience "reboot issues" it would help other like me to understand the problem too, since I don't have the device.

**Comment by yogo1212** [comment]
@chunkeey the uboot file referenced in the original isn't accessible anymore and compex doesn't release it on request.
also, that u-boot doesn't have the cpximg loader which is important for mass-producing hardware using the WPJ428 (no need to boot stock or attach a UART).
without the patch, boards just hang on reboot and they can't be updated either - essentially a brick in remotely managed environments.

i manage installations with several hundred WPJ428. we started using this patch because we experienced the problem described above (no reboot, no sysupgrade). we've encountered two hardware revisions, both of them affected by the issue.

where do you live? i would prefer just supplying you with a board in order to avoid a lengthy back-and-forth.

**Comment by yogo1212** [comment]
I have one variant from the batch that Mehlis used for his initial port (6A03a) and a newer variant (6A04 [1]). Both are affected.
When I take a "fresh" board sold by a reseller in late 2019 and flash it with trunk, it hangs on reboot.
With this patch, it works.
I (very unprofessionally) described my findings in the board support PR in September 2019 (
https://github.com/openwrt/openwrt/pull/787#issuecomment-532410539) back when i thought the SPi flash was causing the problem.

> Compex releaseed back in 2018 an important update for the WPJ428
Do you mean the USB pin?
Boards before and after the USB pin fix don't appear to behave differently in that regard.

Because I wasn't able to aqcuire the u-boot mentioned in the commit, I can't make any tests with that.

Compex themselves are using the GPIO reset - https://github.com/compex-systems/lede/blob/openwrt-18.06/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts#L32.

[1] 04F0214E25D528221959WPJ428HV 6A04PO32256BR04F0214E25D604F0214E25D704F021
4E25D8

**Comment by frolswe** [comment]
The pinctrl / mdio_pins change is needed for me as well. I described why in #2835, basically the bootloader only configures the mdio pins when the interactive u-boot console is activated. Otherwise the pins must be configured by the kernel.

Also, I just found the Compex Lede repository which also contains this change.
https://github.com/compex-systems/lede/blob/openwrt-18.06/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts [hunk 1]
```diff
@@ -32,6 +32,10 @@
 
 		mdio@90000 {
 			status = "okay";
+			pinctrl-0 = <&mdio_pins>;
+			pinctrl-names = "default";
+			reset-gpios = <&tlmm 59 GPIO_ACTIVE_LOW>;
```
Comment (frolswe) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts]: Is GPIO_ACTIVE_LOW correct? The [Compex Lede repository](https://github.com/compex-systems/lede/blob/openwrt-18.06/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts) contains:
`    phy-reset-gpio = <&tlmm 59 0>;`
Which I think means GPIO_ACTIVE_HIGH.

**Review by frolswe — APPROVED** [review]
Looks good to me. Tested 4.19 with bootloader "U-Boot 2012.07 [Chaos Calmer 15.05.1,r35193] (Jul 25 2017 - 11:36:26)".

**Comment by ynezz** [comment]
> Tested 4.19

@frolswe is that Tested-by in https://github.com/ynezz/openwrt/commit/78fe168e5e6830079e49491bb375ce3acccbefcd fine with you?

**Comment by ynezz** [comment]
Thanks, merged into my staging tree https://github.com/ynezz/openwrt/commit/78fe168e5e6830079e49491bb375ce3acccbefcd

**Pull request closed without merging**